File: Arrays/sliding-window-and-two-pointers/PairSum.java
import java.util.Arrays;

public class PairSum {
    public static void main(String[] args) {
        int[] nums = { 3, 7, 8, 12, 23, 38, 49, 51 };
        int target = 10;
        int ans1[] = pairSum(nums, target);
        int ans2[] = pairSumOptimal(nums, target);
        printAns(ans1);
        printAns(ans2);
    }

    // brute force approach
    public static int[] pairSum(int[] nums, int target) {
        int n = nums.length;
        int[] ans = { -1, -1 };
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) {
                if (nums[i] + nums[j] == target) {
                    ans[0] = i;
                    ans[1] = j;
                    return ans;
                }
            }
        }
        return ans;
    }

    // optimal approach
    public static int[] pairSumOptimal(int[] nums, int target) {
        int n = nums.length;
        int start = 0;
        int end = n - 1;
        int[] ans = { -1, -1 };

        while (start < end) {
            if (nums[start] + nums[end] == target) {
                ans[0] = start;
                ans[1] = end;
                return ans;
            } else if (nums[start] + nums[end] < target) {
                start++;
            } else {
                end--;
            }
        }
        return ans;
    }

    public static void printAns(int[] ans) {
        int n = ans.length;
        for (int i = 0; i < 2; i++) {
            System.out.print(ans[i] + " ");
        }
        System.out.println();
    }
}

File: Arrays/sliding-window-and-two-pointers/pair-sum.cpp
#include<iostream>
#include<vector>

using namespace std;


// Brute Force Approach
vector<int>pairSum(vector<int>nums, int target) { // o(n)^2
    int n = nums.size();
    vector<int>ans;
    for (int i = 0; i < n; i++)
    {
        for(int j = i + 1; j < n; j++) {
            if(nums[i] + nums[j] == target) {
                ans.push_back(i);
                ans.push_back(j);
            }
        }
    }
    return ans;
}

// optimal approach - two pointers due to sorted array
vector<int>pairSumOptimal(vector<int>nums, int target) {
    int start = 0; int end = nums.size() - 1;
    vector<int>ans;
    while (start < end) 
    {
        if(nums[start] + nums[end] == target) {
            ans.push_back(start);
            ans.push_back(end);
            return ans;
        } else if(target > nums[start] + nums[end]) {
            start++;
        } else {
            end--;
        }
    }
}

void printAns(vector<int>ans) {
    for(auto value : ans) {
        cout << value << " ";
    }
    cout << endl;
}

int main() {
    vector<int> nums = {2,7,11,15,16};
    int target = 26;
    vector<int>ans = pairSumOptimal(nums, target);
    printAns(ans);
}